identity-iota/src/tangle/client.rs

```diff
@@ -101,6 +111,39 @@ impl Client {
       .map(|message| Receipt::new(self.network.clone(), message))
   }
 
+  /// Publishes arbitrary JSON data to the specified index on the Tangle.
+  /// Retries (promotes or reattaches) the message until it’s included (referenced by a milestone).
+  /// Default interval is 5 seconds and max attempts is 40.
+  pub async fn publish_json_with_retry<T: ToJson>(
+    &self,
+    index: &str,
+    data: &T,
+    interval: Option<u64>,
+    max_attempts: Option<u64>,
+  ) -> Result<Receipt> {
+    let receipt: Receipt = self.publish_json(index, data).await?;
+    let reattached_messages: Vec<(MessageId, Message)> = match self
+      .client
+      .retry_until_included(receipt.message_id(), interval, max_attempts)
+      .await
+    {
+      Ok(reattached_messages) => reattached_messages,
+      Err(e) => match e {
+        IotaClientError::TangleInclusionError(_) => {
+          if self.is_message_included(receipt.message_id()).await? {
+            return Ok(receipt);
+          }
+          return Err(Error::from(e));
+        }
+        _ => return Err(Error::from(e)),
+      },
+    };
+    match reattached_messages.into_iter().next() {
+      Some((_, message)) => Ok(Receipt::new(self.network.clone(), message)),
+      None => Err(Error::MessageInclusionNotChecked),
+    }
```

Comment:
```suggestion
    let receipt: Receipt = self.publish_json(index, data).await?;
    let retry_result = self
      .client
      .retry_until_included(receipt.message_id(), interval, max_attempts)
      .await;

    let reattached_messages: Vec<(MessageId, Message)> = match retry_result {
      Ok(reattached_messages) => reattached_messages,
      Err(inclusion_error @ IotaClientError::TangleInclusionError(_)) => {
        if self.is_message_included(receipt.message_id()).await? {
          return Ok(receipt);
        } else {
          return Err(Error::from(inclusion_error));
        }
      }
      Err(error) => {
        return Err(Error::from(error));
      }
    };

    match reattached_messages.into_iter().next() {
      Some((_, message)) => Ok(Receipt::new(self.network.clone(), message)),
      None => Err(Error::from(IotaClientError::TangleInclusionError(
        receipt.message_id().to_string(),
      ))),
    }
```

Maybe restructure the section a bit:
- Untangle the match statement as much as possible, since that is fairly complex
- I don't think we need a custom error, we can return `TangleInclusionError`, since `MessageInclusionNotChecked` ultimately means the same to the user, i.e. they cannot be sure the message was included.